Add the $previous argument to all exceptions

chinpei215 opened the pull request:
This pull request is the first step to support exception chaining (#9824).

While adding an additional argument to the constructor is usually quite backward compatible, but I would like to target 3.next branch, as this is a little large change.

This is a work in progress, as I am not sure how I should write related test cases. I was going to write test cases for each effected exception class, like the following:
```php
class MissingActionExceptionTest extends TestCase
{

    /**
     * Test the getPrevious() method works.
     *
     * @return void
     */
    public function testGetPrevious()
    {
        $previous = new Exception();
        $exception = new MissingActionException('message', -1, $previous);
        $this->assertSame('message', $exception->getMessage());
        $this->assertSame(-1, $exception->getCode());
        $this->assertSame($previous, $exception->getPrevious());
    }

    /**
     * Test the getPrevious() method works for PHP7 Error
     */
    public function testGetPreviousError() {
        $this->skipIf(version_compare(PHP_VERSION, '7.0.0', '<'));

        $previous = new Error();
        $exception = new MissingActionException('message', -1, $previous);
        $this->assertSame('message', $exception->getMessage());
        $this->assertSame(-1, $exception->getCode());
        $this->assertSame($previous, $exception->getPrevious());
    }
}
```

But should I create something like a `ExceptionsTest` and test all exceptions there instead?

dereuromark commented:
Thanks for doing this!

chinpei215 commented:
Thank you.

I forgot that I had one more question. Should we *overload* the constructor of the exceptions like the follwoing? 
```php
public function __construct($message, $code, $previous);
public function __construct($message, $previous);
```
If it is a sub class of the `HttpException`, the following code doesn't make sense:
```php
catch (Exception $e) {
    throw new NotFoundException('message', 404, $e);
}
```

Or should we create a static factory for the purpose instead?
```php
public static function withPrevious($message, $previous);
```

markstory commented:
> But should I create something like a ExceptionsTest and test all exceptions there instead?

You could create a test case, that uses a data provider to ensure all the known exception classes implement the correct 'interface'.

> I forgot that I had one more question. Should we overload the constructor of the exceptions like the follwoing?

That sounds like it will add complexity that we could avoid. Instead of overloading could use use `null` as a way to avoid duplication around status codes?

```php
catch (Exception $e) {
    throw new NotFoundException('message', null, $e);
}
```

chinpei215 committed:
Add the $previous argument to all exceptions

Add _defaultCode property to Cake\Core\Exception\Exception.
This allows people to pass $previous argument without $code.
Also, now InvalidPrimaryKeyException, RecordNotFoundException,
SocketException, RedirectException and XmlException are subclass of
Cake\Core\Exception\Exception.

chinpei215 commented:
Sorry, I totally forgot about this PR... Now I added test cases.

chinpei215 commented:
Additional notable changes:
- Add _defaultCode property to Cake\Core\Exception\Exception. This allows people to pass $previous argument without $code.
- Now InvalidPrimaryKeyException, RecordNotFoundException, SocketException, RedirectException and XmlException are subclass of Cake\Core\Exception\Exception.

lorenzo commented:
Thanks for doing this @chinpei215 !

markstory commented on tests/TestCase/ExceptionsTest.php:
```diff
@@ -0,0 +1,185 @@
+    public function exceptionProvider()
+    {
+        return [
+            ['Cake\Console\Exception\ConsoleException', 500],
```
Couldn't the default code be read from the class using reflection?

chinpei215 replied:
It could. But then, test cases will pass even if unexpected code is defined in the class.

markstory replied:
Good point.
